src/kudu/consensus/raft_consensus.cc

```diff
@@ -2579,7 +2584,7 @@ void RaftConsensus::DoElectionCallback(ElectionReason reason, const ElectionResu
     LOG_WITH_PREFIX_UNLOCKED(INFO) << "Received " << election_type << " callback for term "
                                    << election_term << " while not running: "
                                    << s.ToString();
-    return;
+    return edcb_(result);
```

Comment:
this election callback will be called with the lock held. 
You should make this callback with no lock held. If this makes it more difficult,
you will somehow have to make it a tail call, like the pattern in.

void LeaderElection::CheckForDecision() 

Alternately you could wrap DoElectionCallback into DoElectionCallbackNested,
which does DoElectionCallback and then fires Nested Callback